src/thumbnailer_slope_optim.cc

```diff
@@ -250,4 +267,80 @@ Thumbnailer::Status Thumbnailer::LossyEncodeNoSlopeOptim(
   return (webp_data->size > 0) ? kOk : kByteBudgetError;
 }
 
+Thumbnailer::Status Thumbnailer::ExtraLossyEncode(WebPData* const webp_data) {
+  // Encode frames following the ascending order of slope between frame's
+  // current quality and quality 100.
+  std::vector<std::pair<float, int>> encoding_order;
+  for (int i = 0; i < frames_.size(); ++i)
+    if (!frames_[i].near_lossless) {
+      float slope;
+      CHECK_THUMBNAILER_STATUS(
+          ComputeSlope(i, frames_[i].final_quality, 100, &slope));
+      encoding_order.push_back(std::make_pair(slope, i));
+    }
+  std::sort(encoding_order.begin(), encoding_order.end(),
+            [](const std::pair<float, int>& x, const std::pair<float, int>& y) {
+              return x.first < y.first;
+            });
+
+  WebPData new_webp_data;
+  WebPDataInit(&new_webp_data);
+  int num_remaining_frames = encoding_order.size();
+
+  while (!encoding_order.empty()) {
+    int min_quality = 100;
+    for (int i = 0; i < num_remaining_frames; ++i) {
+      const int curr_ind = encoding_order[i].second;
+      min_quality = std::min(min_quality, frames_[curr_ind].final_quality + 1);
+    }
+    int max_quality = std::min(min_quality + 30, 100);
+    int final_quality = -1;
+
+    while (min_quality <= max_quality) {
+      int mid_quality = (min_quality + max_quality) / 2;
+      for (int i = 0; i < num_remaining_frames; ++i) {
+        const int curr_ind = encoding_order[i].second;
+        frames_[curr_ind].config.quality =
+            std::max(frames_[curr_ind].final_quality, mid_quality);
+      }
+      CHECK_THUMBNAILER_STATUS(GenerateAnimationNoBudget(&new_webp_data));
+      if (new_webp_data.size <= byte_budget_) {
+        final_quality = mid_quality;
+        WebPDataClear(webp_data);
+        *webp_data = new_webp_data;
+        min_quality = mid_quality + 1;
+      } else {
+        max_quality = mid_quality - 1;
+        WebPDataClear(&new_webp_data);
+      }
+    }
+    if (final_quality != -1) {
+      for (int i = 0; i < num_remaining_frames; ++i) {
+        const int curr_ind = encoding_order[i].second;
+        if (frames_[curr_ind].final_quality < final_quality) {
+          frames_[curr_ind].config.quality = final_quality;
+          frames_[curr_ind].final_quality = final_quality;
+          CHECK_THUMBNAILER_STATUS(
+              GetPictureStats(curr_ind, &frames_[curr_ind].encoded_size,
+                              &frames_[curr_ind].final_psnr));
+        }
+      }
+    } else {
+      break;
+    }
+
+    encoding_order.erase(encoding_order.begin());
+    --num_remaining_frames;
+  }
+
+  std::cerr << "(Final quality, Near-lossless) :" << std::endl;
```

Comment:
Still no logging feature?

Reply:
I am not sure what "logging feature" actually is here. Could you explain more precisely please ?

Reply:
We talked about it during the daily meetings: currently functions always output debugging information (and to `std::cerr` which should be kept for error messages only). It would be cleaner to have them disabled by default. It can be enabled either with a compile flag (something like `#if THUMBNAILER_DEBUG`) or through a binary flag (`-verbose`), up to you. In another PR though.

Reply:
This will be resolved in another PR.